src/language-markdown/whitespace.js

```diff
@@ -0,0 +1,265 @@
+import { getAncestorNode } from "./printer-markdown.js";
+import {
+  KIND_CJK_PUNCTUATION,
+  KIND_CJ_LETTER,
+  KIND_K_LETTER,
+  KIND_NON_CJK,
+  punctuationRegex,
+} from "./utils.js";
+
+const SINGLE_LINE_NODE_TYPES = ["heading", "tableCell", "link", "wikiLink"];
+
+// https://en.wikipedia.org/wiki/Line_breaking_rules_in_East_Asian_languages
+/**
+ * The set of characters that must not immediately precede a line break
+ *
+ * e.g. `"（"`
+ *
+ * - Bad:  `"檜原村（\nひのはらむら）"`
+ * - Good: `"檜原村\n（ひのはらむら）"` or ``"檜原村（ひ\nのはらむら）"`
+ */
+const noBreakAfter = new Set(
+  "$(£¥·'\"〈《「『【〔〖〝﹙﹛＄（［｛￡￥[{‵︴︵︷︹︻︽︿﹁﹃﹏〘｟«"
+);
+/**
+ * The set of characters that must not immediately follow a line break
+ *
+ * e.g. `"）"`
+ *
+ * - Bad:  `"檜原村（ひのはらむら\n）以外には、"`
+ * - Good: `"檜原村（ひのはらむ\nら）以外には、"` or `"檜原村（ひのはらむら）\n以外には、"`
+ */
+const noBreakBefore = new Set(
+  "!%),.:;?]}¢°·'\"†‡›℃∶、。〃〆〕〗〞﹚﹜！＂％＇），．：；？］｝～–—•〉》」︰︱︲︳﹐﹑﹒﹓﹔﹕﹖﹘︶︸︺︼︾﹀﹂﹗｜､』】〙〟｠»ヽヾーァィゥェォッャュョヮヵヶぁぃぅぇぉっゃゅょゎゕゖㇰㇱㇲㇳㇴㇵㇶㇷㇸㇹㇺㇻㇼㇽㇾㇿ々〻‐゠〜～‼⁇⁈⁉・"
+);
+/**
+ * The set of characters whose surrounding newline may be converted to Space
+ *
+ * - ASCII punctuation marks
+ */
+const lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet = new Set(
```

Comment:
Please change "Convert" to "Converts" and remove "SymbolSet".

Reply:
@thorn0 I did